File: scout-ui/lib/nudges.ts
/**
 * Sprint 05 Phase 2 - nudges API helpers.
 * Plain fetch helpers to match the lib/ai-conversations.ts /
 * lib/ai-personality.ts pattern (no react-query).
 */

import { API_BASE_URL } from "./config";
import { del, get, patch, post, put } from "./api";

export type NudgeStatus = "pending" | "delivered" | "suppressed";
export type NudgeSeverity = "low" | "normal" | "high";

export interface NudgeDispatchItem {
  id: string;
  trigger_kind: string;
  trigger_entity_kind: string;
  trigger_entity_id: string | null;
  occurrence_at_utc: string;
  occurrence_local_date: string;
  source_metadata: Record<string, unknown>;
  created_at: string;
}

export interface NudgeDispatch {
  id: string;
  family_member_id: string;
  status: NudgeStatus;
  severity: NudgeSeverity;
  suppressed_reason: string | null;
  deliver_after_utc: string;
  delivered_at_utc: string | null;
  delivered_channels: string[];
  source_count: number;
  body: string | null;
  items: NudgeDispatchItem[];
  created_at: string;
  updated_at: string;
}

export interface QuietHoursConfig {
  start_local_minute: number;
  end_local_minute: number;
  is_default: boolean;
}

export function listMyNudges(limit: number = 20): Promise<NudgeDispatch[]> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  return get(`${API_BASE_URL}/api/nudges/me?${params.toString()}`);
}

export function getFamilyQuietHours(): Promise<QuietHoursConfig> {
  return get(`${API_BASE_URL}/api/admin/family-config/quiet-hours`);
}

export function putFamilyQuietHours(
  start_local_minute: number,
  end_local_minute: number,
): Promise<QuietHoursConfig> {
  return put(`${API_BASE_URL}/api/admin/family-config/quiet-hours`, {
    start_local_minute,
    end_local_minute,
  });
}

// ---------------------------------------------------------------------------
// Sprint 05 Phase 4 - Custom nudge rules (admin)
// ---------------------------------------------------------------------------

export type RuleSeverity = "low" | "normal" | "high";

export interface NudgeRule {
  id: string;
  family_id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  source_kind: "sql_template" | "predicate";
  template_sql: string | null;
  canonical_sql: string | null;
  template_params: Record<string, unknown>;
  trigger_kind: string;
  default_lead_time_minutes: number;
  severity: RuleSeverity;
  created_by_family_member_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface NudgeRuleCreate {
  name: string;
  description?: string;
  source_kind?: "sql_template";
  template_sql: string;
  template_params?: Record<string, unknown>;
  default_lead_time_minutes?: number;
  severity?: RuleSeverity;
  is_active?: boolean;
}

export interface NudgeRulePatch {
  name?: string;
  description?: string | null;
  template_sql?: string;
  template_params?: Record<string, unknown>;
  default_lead_time_minutes?: number;
  severity?: RuleSeverity;
  is_active?: boolean;
}

export interface PreviewCountResponse {
  count: number;
  capped: boolean;
  error: string | null;
}

export function listNudgeRules(): Promise<NudgeRule[]> {
  return get(`${API_BASE_URL}/api/admin/nudges/rules`);
}

export function createNudgeRule(body: NudgeRuleCreate): Promise<NudgeRule> {
  return post(`${API_BASE_URL}/api/admin/nudges/rules`, body);
}

export function patchNudgeRule(
  id: string,
  body: NudgeRulePatch,
): Promise<NudgeRule> {
  return patch(`${API_BASE_URL}/api/admin/nudges/rules/${id}`, body);
}

export function deleteNudgeRule(id: string): Promise<void> {
  return del(`${API_BASE_URL}/api/admin/nudges/rules/${id}`);
}

export function previewRuleCount(id: string): Promise<PreviewCountResponse> {
  return post(`${API_BASE_URL}/api/admin/nudges/rules/${id}/preview-count`, {});
}

File: backend/app/schemas/nudges.py
"""Pydantic read schemas for NudgeDispatch + NudgeDispatchItem.

Sprint 05 Phase 2 — backs GET /api/nudges/me (Recent-nudges UI on
/settings/ai). Read-only; no write schemas yet. Child items are
embedded so the UI does not need a second round-trip to render the
trigger list for a batched dispatch.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field, model_validator


class NudgeDispatchItemRead(BaseModel):
    id: uuid.UUID
    trigger_kind: str
    trigger_entity_kind: str
    trigger_entity_id: uuid.UUID | None
    occurrence_at_utc: datetime
    occurrence_local_date: date
    source_metadata: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class NudgeDispatchRead(BaseModel):
    id: uuid.UUID
    family_member_id: uuid.UUID
    status: str
    severity: str
    suppressed_reason: str | None
    deliver_after_utc: datetime
    delivered_at_utc: datetime | None
    delivered_channels: list[str]
    source_count: int
    body: str | None
    items: list[NudgeDispatchItemRead]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class QuietHoursRead(BaseModel):
    """GET response. Returns the effective window for the family.
    When no row exists, the backend returns the system default
    (22*60=1320 start, 7*60=420 end) so the UI always has a value
    to render."""

    start_local_minute: int
    end_local_minute: int
    is_default: bool  # True when no quiet_hours_family row exists


class QuietHoursUpdate(BaseModel):
    start_local_minute: int = Field(ge=0, lt=1440)
    end_local_minute: int = Field(ge=0, lt=1440)

    @model_validator(mode="after")
    def _start_not_equal_end(self) -> "QuietHoursUpdate":
        if self.start_local_minute == self.end_local_minute:
            raise ValueError("start_local_minute must differ from end_local_minute")
        return self


# Sprint 05 Phase 4 -- admin rule engine


class NudgeRuleRead(BaseModel):
    """GET / list response shape."""

    id: uuid.UUID
    family_id: uuid.UUID
    name: str
    description: str | None
    is_active: bool
    source_kind: str
    template_sql: str | None
    canonical_sql: str | None
    template_params: dict
    trigger_kind: str
    default_lead_time_minutes: int
    severity: str
    created_by_family_member_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class NudgeRuleCreate(BaseModel):
    """POST body. template_sql is required for source_kind='sql_template'
    (the only v1 source_kind). canonical_sql is produced server-side by
    validate_rule_sql; clients do not submit it."""

    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    source_kind: str = Field(default="sql_template", pattern="^sql_template$")
    template_sql: str = Field(min_length=1, max_length=4000)
    template_params: dict = Field(default_factory=dict)
    default_lead_time_minutes: int = Field(default=0, ge=0, le=1440)
    severity: str = Field(default="normal", pattern="^(low|normal|high)$")
    is_active: bool = True

    model_config = {"extra": "forbid"}


class NudgeRulePatch(BaseModel):
    """PATCH body. All fields optional; server-side validation runs
    validate_rule_sql on any template_sql change and re-populates
    canonical_sql."""

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    template_sql: str | None = Field(default=None, min_length=1, max_length=4000)
    template_params: dict | None = None
    default_lead_time_minutes: int | None = Field(default=None, ge=0, le=1440)
    severity: str | None = Field(default=None, pattern="^(low|normal|high)$")
    is_active: bool | None = None

    model_config = {"extra": "forbid"}


class PreviewCountResponse(BaseModel):
    """POST /preview-count response. The count is capped at 200 in the
    rule executor; any higher value is returned as 200 with
    capped=true."""

    count: int
    capped: bool
    error: str | None = None  # populated when the rule's SQL fails to run
